utils/src/main/java/com/cloud/utils/ssh/SshHelper.java

```diff
@@ -1,209 +1,306 @@
+
+            byte[] buffer = new byte[8192];
+            StringBuffer sbResult = new StringBuffer();
+
+            int currentReadBytes = 0;
+            while (true) {
+                throwSshExceptionIfStdoutOrStdeerIsNull(stdout, stderr);
+
+                if ((stdout.available() == 0) && (stderr.available() == 0)) {
+                    int conditions =
+                            sess.waitForCondition(ChannelCondition.STDOUT_DATA | ChannelCondition.STDERR_DATA | ChannelCondition.EOF | ChannelCondition.EXIT_STATUS,
+                                    waitResultTimeoutInMs);
+
+                    throwSshExceptionIfConditionsTimeout(conditions);
+
+                    if ((conditions & ChannelCondition.EXIT_STATUS) != 0) {
+                        break;
+                    }
+
+                    if (canEndTheSshConnection(waitResultTimeoutInMs, sess, conditions)) {
+                        break;
+                    }
+
+                }
+
+                while (stdout.available() > 0) {
+                    currentReadBytes = stdout.read(buffer);
+                    sbResult.append(new String(buffer, 0, currentReadBytes));
+                }
+
+                while (stderr.available() > 0) {
+                    currentReadBytes = stderr.read(buffer);
+                    sbResult.append(new String(buffer, 0, currentReadBytes));
+                }
+            }
+
+            String result = sbResult.toString();
+
+            if (sess.getExitStatus() == null) {
+                //Exit status is NOT available. Returning failure result.
+                s_logger.error(String.format("SSH execution of command %s has no exit status set. Result output: %s", command, result));
+                return new Pair<Boolean, String>(false, result);
+            }
+
+            if (sess.getExitStatus() != null && sess.getExitStatus().intValue() != 0) {
+                s_logger.error(String.format("SSH execution of command %s has an error status code in return. Result output: %s", command, result));
+                return new Pair<Boolean, String>(false, result);
+            }
+
+            return new Pair<Boolean, String>(true, result);
+        } finally {
+            if (sess != null)
+                sess.close();
+
+            if (conn != null)
+                conn.close();
+        }
+    }
+
+    /**
+     * Handles the SSH connection in case of timeout or exit. If the session is with the timeout
+     * condition it throws an exception; if the channel reach an end of file condition but yet does
+     * not send the exit status, then it return true to break the loop; otherwise, it returns false.
+     *
+     * @param waitResultTimeoutInMs
+     * @param sess
+     * @param conditions
+     * @return boolean
+     * @throws SshException
+     */
+    protected static boolean canEndTheSshConnection(int waitResultTimeoutInMs, com.trilead.ssh2.Session sess, int conditions) throws SshException {

```

Comment:
what about an integration test for this one?
integration test != functional tests that test a system functionality.